File: android_app/app/src/main/java/com/oceansentinels/app/presentation/ui/screens/incidents/IncidentDetailScreen.kt
package com.oceansentinels.app.presentation.ui.screens.incidents

import androidx.compose.foundation.layout.*
import androidx.compose.foundation.rememberScrollState
import androidx.compose.foundation.verticalScroll
import androidx.compose.material.icons.Icons
import androidx.compose.material.icons.filled.*
import androidx.compose.material3.*
import androidx.compose.runtime.*
import androidx.compose.ui.Alignment
import androidx.compose.ui.Modifier
import androidx.compose.ui.text.font.FontWeight
import androidx.compose.ui.unit.dp
import androidx.hilt.navigation.compose.hiltViewModel
import coil.compose.AsyncImage
import com.oceansentinels.app.domain.model.IncidentStatus
import com.oceansentinels.app.domain.model.UserRole
import com.oceansentinels.app.domain.model.UserSession
import com.oceansentinels.app.presentation.ui.components.*
import com.oceansentinels.app.presentation.ui.theme.OceanColors
import com.oceansentinels.app.presentation.viewmodel.AuthViewModel
import com.oceansentinels.app.presentation.viewmodel.IncidentViewModel
import java.text.SimpleDateFormat
import java.util.*

/**
 * Incident detail screen
 */
@OptIn(ExperimentalMaterial3Api::class)
@Composable
fun IncidentDetailScreen(
    incidentId: Int,
    onNavigateBack: () -> Unit,
    onNavigateToMap: (Double, Double) -> Unit,
    incidentViewModel: IncidentViewModel = hiltViewModel(),
    authViewModel: AuthViewModel = hiltViewModel()
) {
    val incident by incidentViewModel.selectedIncident.collectAsState()
    val isLoading by incidentViewModel.isLoading.collectAsState()
    val error by incidentViewModel.error.collectAsState()
    val userSession by authViewModel.userSession.collectAsState()
    
    val currentUser = (userSession as? UserSession.Authenticated)?.user
    val canTakeAction = currentUser?.role in listOf(UserRole.ADMIN, UserRole.RESCUE_TEAM, UserRole.AUTHORITY)
    
    // Load incident details
    LaunchedEffect(incidentId) {
        incidentViewModel.getIncident(incidentId)
    }
    
    val scrollState = rememberScrollState()
    
    Scaffold(
        topBar = {
            TopAppBar(
                title = { Text("Incident Details") },
                navigationIcon = {
                    IconButton(onClick = onNavigateBack) {
                        Icon(Icons.Default.ArrowBack, contentDescription = "Back")
                    }
                },
                actions = {
                    incident?.let {
                        IconButton(onClick = { onNavigateToMap(it.latitude ?: 0.0, it.longitude ?: 0.0) }) {
                            Icon(Icons.Default.Map, contentDescription = "View on Map")
                        }
                    }
                }
            )
        }
    ) { paddingValues ->
        Box(
            modifier = Modifier
                .fillMaxSize()
                .padding(paddingValues)
        ) {
            when {
                isLoading && incident == null -> {
                    Box(
                        modifier = Modifier.fillMaxSize(),
                        contentAlignment = Alignment.Center
                    ) {
                        CircularProgressIndicator()
                    }
                }
                
                error != null && incident == null -> {
                    EmptyStateCard(
                        title = "Error Loading Incident",
                        description = error ?: "Something went wrong",
                        icon = Icons.Default.Error,
                        modifier = Modifier
                            .fillMaxWidth()
                            .padding(16.dp)
                    ) {
                        OceanPrimaryButton(
                            text = "Try Again",
                            onClick = { incidentViewModel.getIncident(incidentId) }
                        )
                    }
                }
                
                incident != null -> {
                    Column(
                        modifier = Modifier
                            .fillMaxSize()
                            .verticalScroll(scrollState)
                            .padding(16.dp)
                    ) {
                        val inc = incident!!
                        
                        // Reference ID
                        Card(
                            modifier = Modifier.fillMaxWidth(),
                            colors = CardDefaults.cardColors(
                                containerColor = OceanColors.Primary.copy(alpha = 0.1f)
                            )
                        ) {
                            Row(
                                modifier = Modifier
                                    .padding(16.dp)
                                    .fillMaxWidth(),
                                horizontalArrangement = Arrangement.SpaceBetween,
                                verticalAlignment = Alignment.CenterVertically
                            ) {
                                Column {
                                    Text(
                                        text = "Reference ID",
                                        style = MaterialTheme.typography.labelMedium,
                                        color = MaterialTheme.colorScheme.onSurfaceVariant
                                    )
                                    Text(
                                        text = inc.referenceId,
                                        style = MaterialTheme.typography.titleMedium,
                                        fontWeight = FontWeight.Bold,
                                        color = OceanColors.Primary
                                    )
                                }
                                StatusBadge(status = inc.status)
                            }
                        }
                        
                        Spacer(modifier = Modifier.height(16.dp))
                        
                        // Title and Description
                        Card(
                            modifier = Modifier.fillMaxWidth()
                        ) {
                            Column(
                                modifier = Modifier.padding(16.dp)
                            ) {
                                Text(
                                    text = inc.hazardType.displayName,
                                    style = MaterialTheme.typography.headlineSmall,
                                    fontWeight = FontWeight.Bold
                                )
                                
                                Spacer(modifier = Modifier.height(12.dp))
                                
                                Row(
                                    horizontalArrangement = Arrangement.spacedBy(8.dp)
                                ) {
                                    HazardTypeBadge(hazardType = inc.hazardType)
                                    UrgencyBadge(urgency = inc.urgency)
                                }
                                
                                Spacer(modifier = Modifier.height(16.dp))
                                
                                Text(
                                    text = "Description",
                                    style = MaterialTheme.typography.titleSmall,
                                    fontWeight = FontWeight.SemiBold
                                )
                                Spacer(modifier = Modifier.height(4.dp))
                                Text(
                                    text = inc.description,
                                    style = MaterialTheme.typography.bodyMedium,
                                    color = MaterialTheme.colorScheme.onSurfaceVariant
                                )
                            }
                        }
                        
                        Spacer(modifier = Modifier.height(16.dp))
                        
                        // Location
                        Card(
                            modifier = Modifier.fillMaxWidth()
                        ) {
                            Column(
                                modifier = Modifier.padding(16.dp)
                            ) {
                                Row(
                                    verticalAlignment = Alignment.CenterVertically
                                ) {
                                    Icon(
                                        imageVector = Icons.Default.LocationOn,
                                        contentDescription = null,
                                        tint = OceanColors.Primary
                                    )
                                    Spacer(modifier = Modifier.width(8.dp))
                                    Text(
                                        text = "Location",
                                        style = MaterialTheme.typography.titleMedium,
                                        fontWeight = FontWeight.SemiBold
                                    )
                                }
                                
                                Spacer(modifier = Modifier.height(12.dp))
                                
                                Text(
                                    text = inc.location,
                                    style = MaterialTheme.typography.bodyMedium
                                )
                                Spacer(modifier = Modifier.height(8.dp))
                                
                                Row(
                                    modifier = Modifier.fillMaxWidth(),
                                    horizontalArrangement = Arrangement.spacedBy(16.dp)
                                ) {
                                    Column {
                                        Text(
                                            text = "Latitude",
                                            style = MaterialTheme.typography.labelSmall,
                                            color = MaterialTheme.colorScheme.onSurfaceVariant
                                        )
                                        Text(
                                            text = "%.6f".format(inc.latitude ?: 0.0),
                                            style = MaterialTheme.typography.bodyMedium,
                                            fontWeight = FontWeight.Medium
                                        )
                                    }
                                    Column {
                                        Text(
                                            text = "Longitude",
                                            style = MaterialTheme.typography.labelSmall,
                                            color = MaterialTheme.colorScheme.onSurfaceVariant
                                        )
                                        Text(
                                            text = "%.6f".format(inc.longitude ?: 0.0),
                                            style = MaterialTheme.typography.bodyMedium,
                                            fontWeight = FontWeight.Medium
                                        )
                                    }
                                }
                                
                                Spacer(modifier = Modifier.height(12.dp))
                                
                                OutlinedButton(
                                    onClick = { onNavigateToMap(inc.latitude ?: 0.0, inc.longitude ?: 0.0) },
                                    modifier = Modifier.fillMaxWidth()
                                ) {
                                    Icon(Icons.Default.Map, contentDescription = null)
                                    Spacer(modifier = Modifier.width(8.dp))
                                    Text("View on Map")
                                }
                            }
                        }
                        
                        Spacer(modifier = Modifier.height(16.dp))
                        
                        // Image if available
                        inc.photoUrl?.let { url ->
                            Card(
                                modifier = Modifier.fillMaxWidth()
                            ) {
                                Column(
                                    modifier = Modifier.padding(16.dp)
                                ) {
                                    Row(
                                        verticalAlignment = Alignment.CenterVertically
                                    ) {
                                        Icon(
                                            imageVector = Icons.Default.Image,
                                            contentDescription = null,
                                            tint = OceanColors.Primary
                                        )
                                        Spacer(modifier = Modifier.width(8.dp))
                                        Text(
                                            text = "Photo Evidence",
                                            style = MaterialTheme.typography.titleMedium,
                                            fontWeight = FontWeight.SemiBold
                                        )
                                    }
                                    
                                    Spacer(modifier = Modifier.height(12.dp))
                                    
                                    AsyncImage(
                                        model = url,
                                        contentDescription = "Incident photo",
                                        modifier = Modifier
                                            .fillMaxWidth()
                                            .height(200.dp)
                                    )
                                }
                            }
                            
                            Spacer(modifier = Modifier.height(16.dp))
                        }
                        
                        // Timeline/Timestamps
                        Card(
                            modifier = Modifier.fillMaxWidth()
                        ) {
                            Column(
                                modifier = Modifier.padding(16.dp)
                            ) {
                                Row(
                                    verticalAlignment = Alignment.CenterVertically
                                ) {
                                    Icon(
                                        imageVector = Icons.Default.Timeline,
                                        contentDescription = null,
                                        tint = OceanColors.Primary
                                    )
                                    Spacer(modifier = Modifier.width(8.dp))
                                    Text(
                                        text = "Timeline",
                                        style = MaterialTheme.typography.titleMedium,
                                        fontWeight = FontWeight.SemiBold
                                    )
                                }
                                
                                Spacer(modifier = Modifier.height(12.dp))
                                
                                TimelineItem(
                                    title = "Reported",
                                    date = inc.createdAt,
                                    isCompleted = true
                                )
                                
                                TimelineItem(
                                    title = "Verified",
                                    date = inc.verifiedAt,
                                    isCompleted = inc.status in listOf(
                                        IncidentStatus.VERIFIED,
                                        IncidentStatus.IN_PROGRESS,
                                        IncidentStatus.RESOLVED
                                    )
                                )
                                
                                TimelineItem(
                                    title = "Response Deployed",
                                    date = inc.updatedAt,
                                    isCompleted = inc.status in listOf(
                                        IncidentStatus.IN_PROGRESS,
                                        IncidentStatus.RESOLVED
                                    )
                                )
                                
                                TimelineItem(
                                    title = "Resolved",
                                    date = inc.resolvedAt,
                                    isCompleted = inc.status == IncidentStatus.RESOLVED,
                                    isLast = true
                                )
                            }
                        }
                        
                        // Actions for authorized users
                        if (canTakeAction && inc.status != IncidentStatus.RESOLVED) {
                            Spacer(modifier = Modifier.height(24.dp))
                            
                            Text(
                                text = "Actions",
                                style = MaterialTheme.typography.titleMedium,
                                fontWeight = FontWeight.SemiBold
                            )
                            
                            Spacer(modifier = Modifier.height(12.dp))
                            
                            when (inc.status) {
                                IncidentStatus.PENDING -> {
                                    OceanPrimaryButton(
                                        text = "Verify Incident",
                                        onClick = { incidentViewModel.verifyIncident(inc.id) },
                                        modifier = Modifier.fillMaxWidth(),
                                        icon = Icons.Default.Verified,
                                        isLoading = isLoading
                                    )
                                }
                                IncidentStatus.VERIFIED -> {
                                    OceanPrimaryButton(
                                        text = "Deploy Response",
                                        onClick = { incidentViewModel.deployResponse(inc.id) },
                                        modifier = Modifier.fillMaxWidth(),
                                        icon = Icons.Default.LocalShipping,
                                        isLoading = isLoading
                                    )
                                }
                                IncidentStatus.IN_PROGRESS -> {
                                    OceanPrimaryButton(
                                        text = "Mark as Resolved",
                                        onClick = { incidentViewModel.resolveIncident(inc.id) },
                                        modifier = Modifier.fillMaxWidth(),
                                        icon = Icons.Default.CheckCircle,
                                        isLoading = isLoading
                                    )
                                }
                                else -> {}
                            }
                        }
                        
                        Spacer(modifier = Modifier.height(32.dp))
                    }
                }
            }
        }
    }
}

@Composable
private fun TimelineItem(
    title: String,
    date: java.time.LocalDateTime?,
    isCompleted: Boolean,
    isLast: Boolean = false
) {
    val dateFormatter = java.time.format.DateTimeFormatter.ofPattern("MMM d, yyyy 'at' h:mm a")
    
    Row(
        modifier = Modifier.height(IntrinsicSize.Min)
    ) {
        // Timeline indicator
        Column(
            horizontalAlignment = Alignment.CenterHorizontally
        ) {
            Surface(
                shape = MaterialTheme.shapes.extraSmall,
                color = if (isCompleted) OceanColors.Success else MaterialTheme.colorScheme.outlineVariant,
                modifier = Modifier.size(12.dp)
            ) {}
            
            if (!isLast) {
                Box(
                    modifier = Modifier
                        .width(2.dp)
                        .weight(1f)
                        .padding(vertical = 4.dp)
                ) {
                    Divider(
                        modifier = Modifier.fillMaxHeight(),
                        color = if (isCompleted) OceanColors.Success else MaterialTheme.colorScheme.outlineVariant
                    )
                }
            }
        }
        
        Spacer(modifier = Modifier.width(12.dp))
        
        Column(
            modifier = Modifier.padding(bottom = if (isLast) 0.dp else 16.dp)
        ) {
            Text(
                text = title,
                style = MaterialTheme.typography.bodyMedium,
                fontWeight = FontWeight.Medium,
                color = if (isCompleted) MaterialTheme.colorScheme.onSurface else MaterialTheme.colorScheme.onSurfaceVariant
            )
            if (date != null) {
                Text(
                    text = date.format(dateFormatter),
                    style = MaterialTheme.typography.bodySmall,
                    color = MaterialTheme.colorScheme.onSurfaceVariant
                )
            } else if (!isCompleted) {
                Text(
                    text = "Pending",
                    style = MaterialTheme.typography.bodySmall,
                    color = MaterialTheme.colorScheme.onSurfaceVariant.copy(alpha = 0.6f)
                )
            }
        }
    }
}

File: IOS/OceanSentinels/Presentation/UI/Screens/Incidents/IncidentDetailScreen.swift
import SwiftUI

// MARK: - IncidentDetailScreen

struct IncidentDetailScreen: View {
    let incidentId: Int

    @Environment(IncidentViewModel.self) private var viewModel
    @Environment(AuthViewModel.self) private var authViewModel
    @Environment(\.dismiss) private var dismiss

    private var currentUser: User? {
        if case .authenticated(let user, _) = authViewModel.userSession { return user }
        return nil
    }
    private var canTakeAction: Bool {
        guard let role = currentUser?.role else { return false }
        return role == .admin || role == .rescueTeam || role == .authority
    }

    var body: some View {
        Group {
            if viewModel.isLoading && viewModel.selectedIncident == nil {
                ProgressView().frame(maxWidth: .infinity, maxHeight: .infinity)
            } else if let error = viewModel.error, viewModel.selectedIncident == nil {
                ContentUnavailableView {
                    Label("Error", systemImage: "exclamationmark.triangle")
                } description: {
                    Text(error)
                } actions: {
                    Button("Try Again") { viewModel.getIncident(id: incidentId) }
                }
            } else if let incident = viewModel.selectedIncident {
                incidentContent(incident)
            }
        }
        .navigationTitle("Incident Details")
        .navigationBarTitleDisplayMode(.inline)
        .task { viewModel.getIncident(id: incidentId) }
    }

    @ViewBuilder
    private func incidentContent(_ inc: Incident) -> some View {
        ScrollView {
            VStack(spacing: 16) {
                // Reference ID card
                HStack {
                    VStack(alignment: .leading, spacing: 4) {
                        Text("Reference ID").font(.caption).foregroundStyle(.secondary)
                        Text(inc.referenceId).font(.headline).foregroundStyle(Color.oceanPrimary)
                    }
                    Spacer()
                    StatusBadge(status: inc.status)
                }
                .padding(16)
                .background(Color.oceanPrimary.opacity(0.1), in: RoundedRectangle(cornerRadius: 12))

                // Title & Description
                VStack(alignment: .leading, spacing: 12) {
                    Text(inc.hazardType.displayName).font(.title3.bold())
                    HStack(spacing: 8) {
                        HazardTypeBadge(hazardType: inc.hazardType)
                        UrgencyBadge(urgency: inc.urgency)
                    }
                    Divider()
                    Text("Description").font(.subheadline.weight(.semibold))
                    Text(inc.description).font(.body).foregroundStyle(.secondary)
                }
                .padding(16)
                .background(Color(.secondarySystemBackground), in: RoundedRectangle(cornerRadius: 12))

                // Location
                VStack(alignment: .leading, spacing: 12) {
                    Label("Location", systemImage: "location.fill")
                        .font(.headline)
                        .foregroundStyle(Color.oceanPrimary)
                    Text(inc.location).font(.body)
                    HStack(spacing: 24) {
                        VStack(alignment: .leading) {
                            Text("Latitude").font(.caption).foregroundStyle(.secondary)
                            Text(String(format: "%.6f", inc.latitude ?? 0)).font(.body.weight(.medium))
                        }
                        VStack(alignment: .leading) {
                            Text("Longitude").font(.caption).foregroundStyle(.secondary)
                            Text(String(format: "%.6f", inc.longitude ?? 0)).font(.body.weight(.medium))
                        }
                    }
                }
                .padding(16)
                .background(Color(.secondarySystemBackground), in: RoundedRectangle(cornerRadius: 12))

                // Photo
                if let photoUrl = inc.photoUrl, let url = URL(string: photoUrl) {
                    VStack(alignment: .leading, spacing: 12) {
                        Label("Photo Evidence", systemImage: "photo")
                            .font(.headline)
                            .foregroundStyle(Color.oceanPrimary)
                        AsyncImage(url: url) { phase in
                            switch phase {
                            case .success(let image):
                                image.resizable().scaledToFit().frame(maxHeight: 200).clipShape(RoundedRectangle(cornerRadius: 8))
                            case .failure:
                                Image(systemName: "photo.badge.exclamationmark").font(.largeTitle).foregroundStyle(.secondary)
                            default:
                                ProgressView()
                            }
                        }
                    }
                    .padding(16)
                    .background(Color(.secondarySystemBackground), in: RoundedRectangle(cornerRadius: 12))
                }

                // Timeline
                VStack(alignment: .leading, spacing: 12) {
                    Label("Timeline", systemImage: "chart.line.uptrend.xyaxis")
                        .font(.headline)
                        .foregroundStyle(Color.oceanPrimary)
                    timelineItem(title: "Reported", date: inc.createdAt, isCompleted: true)
                    timelineItem(title: "Verified", date: inc.verifiedAt, isCompleted: [.verified, .inProgress, .resolved].contains(inc.status))
                    timelineItem(title: "Response Deployed", date: inc.updatedAt, isCompleted: [.inProgress, .resolved].contains(inc.status))
                    timelineItem(title: "Resolved", date: inc.resolvedAt, isCompleted: inc.status == .resolved, isLast: true)
                }
                .padding(16)
                .background(Color(.secondarySystemBackground), in: RoundedRectangle(cornerRadius: 12))

                // Actions
                if canTakeAction && inc.status != .resolved {
                    VStack(spacing: 12) {
                        Text("Actions").font(.headline)
                        switch inc.status {
                        case .pending:
                            OceanPrimaryButton(text: "Verify Incident", action: { viewModel.verifyIncident(id: inc.id) }, isLoading: viewModel.isLoading, icon: "checkmark.seal.fill")
                        case .verified:
                            OceanPrimaryButton(text: "Deploy Response", action: { viewModel.deployResponse(id: inc.id) }, isLoading: viewModel.isLoading, icon: "shippingbox.fill")
                        case .inProgress:
                            OceanPrimaryButton(text: "Mark as Resolved", action: { viewModel.resolveIncident(id: inc.id) }, isLoading: viewModel.isLoading, icon: "checkmark.circle.fill")
                        default:
                            EmptyView()
                        }
                    }
                }
            }
            .padding(16)
        }
    }

    @ViewBuilder
    private func timelineItem(title: String, date: Date?, isCompleted: Bool, isLast: Bool = false) -> some View {
        HStack(alignment: .top, spacing: 12) {
            VStack(spacing: 4) {
                Circle()
                    .fill(isCompleted ? Color.oceanSuccess : Color(.systemGray4))
                    .frame(width: 12, height: 12)
                if !isLast {
                    Rectangle()
                        .fill(isCompleted ? Color.oceanSuccess : Color(.systemGray4))
                        .frame(width: 2, height: 24)
                }
            }
            VStack(alignment: .leading, spacing: 2) {
                Text(title).font(.subheadline.weight(.medium)).foregroundStyle(isCompleted ? .primary : .secondary)
                if let date {
                    Text(date, format: .dateTime.month(.abbreviated).day().year().hour().minute())
                        .font(.caption).foregroundStyle(.secondary)
                } else if !isCompleted {
                    Text("Pending").font(.caption).foregroundStyle(.tertiary)
                }
            }
        }
    }
}
